OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs

```diff
@@ -0,0 +1,480 @@
+        IEUSignCP.SaveCertificates(
+            File.ReadAllBytes(eUSignConfig.CA.CertificatesPath));
+
+        // Створення контексту бібліотеки
+        IEUSignCP.CtxCreate(out context);
+    }
+
+    /// <summary>
+    /// This code is from IIT Library usage example with code styling changes.
+    /// Original code:
+    /// foreach (var ca in CAs)
+    /// {
+    ///     foreach (var issuer in ca.issuerCNs)
+    ///     {
+    ///         if (caSubjectCN == issuer)
+    ///         {
+    ///             return ca;
+    ///         }
+    ///     }
+    /// }
+    ///
+    /// return null;
+    /// Returns CA parameters by subject name.
+    /// </summary>
+    /// <param name="caSubjectCN">Required CS Subject Common Name.</param>
+    /// <returns>A <see cref="CASettings"/> with required CA parameters.</returns>
+    private CASettings GetCA(string caSubjectCN) => CAs.FirstOrDefault(ca => ca.issuerCNs.Any(cn => cn == caSubjectCN));
+
+    /// <summary>
+    /// This code is from IIT Library usage example with minimal code styling changes.
+    /// </summary>
+    private bool IsPrivateKeyLoaded()
+    {
+        return pkContext != IntPtr.Zero;
+    }
+
+    /// <summary>
+    /// This code is from IIT Library usage example with minimal code styling changes.
+    /// </summary>
+    private void ReadPrivateKey(
+        string pkFile,
+        string pkTypeName,
+        string pkDeviceName,
+        string password,
+        string jksAlias,
+        string[] certsFiles,
+        string caIssuerCN)
+    {
+        byte[] pKey = null;
+        List<byte[]> certs = new List<byte[]>();
+        bool fromFile = !string.IsNullOrEmpty(pkFile);
+        IEUSignCP.EU_KEY_MEDIA keyMedia = new IEUSignCP.EU_KEY_MEDIA(0, 0, string.Empty);
+        CASettings ca = null;
+        IEUSignCP.EU_CERT_OWNER_INFO info;
+        IEUSignCP.EU_CERT_INFO_EX infoEx;
+
+        try
+        {
+            // Зчитування сертифікатів ос. ключа, з файлової системи (якщо задані)
+            if (certsFiles != null && certsFiles.Length > 0)
+            {
+                foreach (var file in certsFiles)
+                {
+                    if (!File.Exists(file))
+                    {
+                        throw new Exception($"The certificate file is missing: {file}");
+                    }
+
+                    certs.Add(File.ReadAllBytes(file));
+                }
+            }
+
+            if (fromFile)
+            {
+                // Зчитування ос. ключа з файлу до байтового масиву
+                if (!File.Exists(pkFile))
+                {
+                    throw new Exception($"The private key file is missing: {pkFile}");
+                }
+
+                pKey = File.ReadAllBytes(pkFile);
+                if (!string.IsNullOrEmpty(jksAlias))
+                {
+                    // Пошук відповідного ос. ключа в контейнері JKS
+                    IEUSignCP.GetJKSPrivateKey(pKey, jksAlias, out pKey, out var certificates);
```

Comment:
@h4wk13 according to docs certain certificate authorities have certificates as a container for set of different keys instead of being plain certificate. So we read the certificate and if it is a container (we specify the alias to search) - we use it and the key alias to find the actual certificate in the container

Reply:
Sounds promising...